### PR TITLE
bulk delete

### PR DESCRIPTION
**Pull request by ereteog**
> related #https://github.com/threatgrid/iroh/issues/2250
> related #https://github.com/threatgrid/iroh/issues/4960

This PR implements bulk delete (and few cleaning).
Don't be afraid by the size, 1/3 is documentation.
For a quicker review, I recommend this order
* ES store
  * ctia.stores.es.crud
  * ctia.stores.es.crud-test
* flow
  * ctia.flow.crud
  * ctia.flow.crud-test
* bulk core
  * ctia.bulk.core
  * ctia.bulk.core-test
* bulk route
  * ctia.bulk.route
  * ctia.bulk.route-test

<a name="qa">[§](#qa)</a> QA
============================
### Standard scenario

1) Bulk create the following entities and note the all the created ids
[bulk-delete-test.json.zip](https://github.com/threatgrid/ctia/files/6755357/bulk-delete-test.json.zip)
2) Bulk delete 2 of the created indicators (`indicator-id-1`, `indicator-id-2`) and one the created malwares (`malware-id-1`) with the following body:
``` javascript
{
   "indicators":[
      "indicator-id-1",
      "indicator-id-2"
   ],
   "malware":[
      "malware-id"
   ]
}
```
check that the body of the response is
``` javascript
{
   "indicators":{
      "deleted":[
         "indicator-id-1",
         "indicator-id-2"
      ]
   },
   "malware":{
      "deleted":[
         "malware-id-1"
      ]
   }
}
```
3) try to retrieve `indicator-id-1`, `indicator-id-2` and `malware-id-1` and check that you get a 404 for each of them.
4) try to delete all previously created indicators (including those already deleted)
``` javascript
{
   "indicators":[
      "indicator-id-1",
      "indicator-id-2",
      "indicator-id-3",
      "indicator-id-4"
   ],
   "malware":[
      "malware-id-1",
      "malware-id-2",
      "malware-id-3"
   ]
}
```
check that the body of the response is
``` javascript
{
   "indicators":{
      "deleted":[
         "indicator-id-3",
         "indicator-id-4"
      ],
      "errors":{
         "not-found":[
            "indicator-id-1",
            "indicator-id-2"
         ]
      }
   },
   "malware":{
      "deleted":[
         "malware-id-2",
         "malware-id-3"
      ],
      "errors":{
         "not-found":[
            "malware-id-1"
         ]
      }
   }
}
```

### Access right tests 
IN public intel
1) with user1 from group1 create 2 judgements with TLP amber and 1 judgement with tlp green:
``` javascript
{
        "judgements": [
            {
                "valid_time": {
                    "start_time": "2016-02-11T00:40:48.212Z",
                    "end_time": "2016-07-11T00:40:48.212Z"
                },
                "observable": {
                    "value": "10.0.0.0",
                    "type": "ip"
                },
                "disposition": 2,
                "source": "test",
                "priority": 100,
                "severity": "High",
                "confidence": "Low",
                "tlp": "amber"
            },
            {
                "valid_time": {
                    "start_time": "2016-02-11T00:40:48.212Z",
                    "end_time": "2016-07-11T00:40:48.212Z"
                },
                "observable": {
                    "value": "10.0.0.1",
                    "type": "ip"
                },
                "disposition": 2,
                "source": "test",
                "priority": 100,
                "severity": "High",
                "confidence": "Low",
                "tlp": "amber"
            },
            {
                "valid_time": {
                    "start_time": "2016-02-11T00:40:48.212Z",
                    "end_time": "2016-07-11T00:40:48.212Z"
                },
                "observable": {
                    "value": "10.0.0.6",
                    "type": "ip"
                },
                "disposition": 2,
                "source": "test",
                "priority": 100,
                "severity": "High",
                "confidence": "Low",
                "tlp": "green"
            }
        ]
}
```
2) with user2 from group2 with write access, bulk delete these entities, they should all not be found since the user must not have the visibility on these entities:
``` javascript
{
   "judgement":{
      "errors":{
         "not-found":[
            "jugment-id-1",
            "jugment-id-2"
         ],
         "forbidden": [
             "jugment-id-3"
          ]
      }
   }
}
```
Check, with user1, that these entities are still retrievable.
 
3) with user3 from group1 bulk delete these entities :
``` javascript
{
   "judgement":{
      "deleted": [
            "jugment-id-1",
            "jugment-id-2",
            "jugment-id-3"
      ]
   }
}
```
check with user1 that these entities are actually deleted.

<a name="release-notes">[§](#release-notes)</a> Release Notes
=============================================================

```
intern: Bulk delete route in CTIA.
```

### CODE REVIEW

File: src/ctia/stores/es/crud.clj [hunk 1]
```diff
@@ -91,23 +93,43 @@
...
+ documents from an alias that points to multiple indices.
+It returns the documents with full hits meta data including the real index in which is stored the document."
+  ([conn-state ids es-params batch-size]
+   (lazy-get-docs-with-indices conn-state ids es-params batch-size get-docs-with-indices))
+  ([conn-state
+    ids
+    es-params
+    batch-size
+    docs-by-ids]
+   (when (seq ids)
```
Comment (vemv) [inline src/ctia/stores/es/crud.clj:124]: maybe `(s/pred seq)` the `ids` instead, this way one gets faster/clearer feedback than a nil result

File: src/ctia/stores/es/crud.clj [hunk 2]
```diff
@@ -201,8 +237,49 @@ It returns the documents with full hits meta data including the real index in wh
...
+   ident
+   es-params]
+  (let [docs-with-indices (get-docs-with-indices conn-state ids {})
+        {authorized true unauthorized false}
+        (group-by #(allow-write? (:_source %) ident)
+                  docs-with-indices)
+        missing (set/difference (set ids) (map :_id docs-with-indices))
+        prepared-docs (map #(select-keys % [:_index :_type :_id])
+                           authorized)]
+    (try
```
Comment (vemv) [inline src/ctia/stores/es/crud.clj:259]: Looks like the only part that can fail is `ductile.doc/bulk-delete-docs`

It tends to be better to thinly wrap in `try` just that part. That way, trivial programming errors aren't swallowed, which  makes things harder to debug

File: test/ctia/stores/es/crud_test.clj [hunk 3]
```diff
@@ -568,3 +594,151 @@
...
+             total 200
+             sighting-ids (map #(str "sighting-"  %) (range total))
+             mk-title #(str "title "%)
+             sightings (map #(assoc base-sighting
+                                    :id %
+                                    :title (mk-title %))
+                            sighting-ids)
+             ;; drop some docs to check that we only match requeted ids
+             tested-sample (drop (/ total 10)
+                                 (shuffle sightings))
```
Comment (vemv) [inline test/ctia/stores/es/crud_test.clj:617]: Using `shuffle` in tests seems delicate since tests would ideally would always be deterministic/reproducible.

Is there some alternative?

File: src/ctia/stores/es/crud.clj (hunk 1 above)
Comment (ereteog) [reply src/ctia/stores/es/crud.clj:124]: That's the end condition of the (lazy) recursion. 
BTW, if `ids` is also lazily generated (not likely but still), adding a schema would make  it eagerly consumed in order to check the schema. It's also probable that it could also break the lazyness of this function (I don't think so but it worths checking 😀, I will do so when I will be back from my ankle check).

File: test/ctia/stores/es/crud_test.clj [hunk 4]
```diff
@@ -568,3 +594,151 @@
...
+             base-sighting {:tlp "green"
+                            :groups ["group1"]}
+             total 200
+             sighting-ids (map #(str "sighting-"  %) (range total))
+             mk-title #(str "title "%)
+             sightings (map #(assoc base-sighting
+                                    :id %
+                                    :title (mk-title %))
+                            sighting-ids)
+             ;; drop some docs to check that we only match requeted ids
```
Comment (agzam) [inline test/ctia/stores/es/crud_test.clj:615]: typo

File: src/ctia/stores/es/crud.clj [hunk 5]
```diff
@@ -92,23 +87,43 @@
...
                                (assoc (make-es-read-params es-params)
+                                      :limit (count ids)
                                       :full-hits?
                                       true))]
     (:data res)))
 
+(defn lazy-get-docs-with-indices
+  "Retrieves a documents from a search \"ids\" query. It enables to retrieves
+ documents from an alias that points to multiple indices.
+It returns the documents with full hits meta data including the real index in which is stored the document."
```
Comment (agzam) [inline src/ctia/stores/es/crud.clj:110]: This docstring sounds like a hip-hop song :)

File: test/ctia/flows/crud_test.clj [hunk 6]
```diff
@@ -122,3 +126,113 @@
...
+                                                                some?
+                                                                (array-map %)))
+                                                     entity-ids)]
+                           (is (= expected
+                                  (into {} res)))
+                           (is (= expected deleted-events?)
+                               (str "flow shall return " expected))))]
+       (delete-flow {"sighting-1" true
+                     "sighting-2" true})
+       (delete-flow {"sighting-3" true "missing" false})))))
```
[the remark this answers is not in the document]
Comment (yogsototh) [reply test/ctia/flows/crud_test.clj:238]: Another comment it would probably be a good idea to test in a concurrent environment.

Like receiving two delete-flow on a non disjoint set of entities. This way you should be able to test that.

```
(concurrently
   (delete-flow sightings-1 ... sightings-10)
   (delete-flow sightings-5 .. sighting-15)
   (delete-flow sightings-3 .. sighting-8)
   (delete-flow sightings-10 .. sighting-20))
```

would deletes from sighting-1 to sighting-20. Even if for some call not all the entities asked to be removed will be removed during this specific call.

File: test/ctia/stores/es/crud_test.clj [hunk 7]
```diff
@@ -581,3 +607,105 @@
...
+                                    :groups groups))
+             ident1 {:login "john-doe"
+                     :groups ["group1"]}
+             ident2 {:login "jane-doe"
+                     :groups ["group2"]}
+
+             [docs-1 docs-2 docs-3] (partition-all (/ total 3) sightings)
+             docs1-ident1 (map (partial with-owner ident1) docs-1)
+             docs2-ident2 (map (partial with-owner ident2) docs-2)
+             docs3-ident1 (map (partial with-owner ident1) docs-3)
```
Comment (yogsototh) [inline test/ctia/stores/es/crud_test.clj:683]:
```suggestion
             ;; document created and owned
             docs1-ident1 (map (partial with-owner ident1) docs-1)
             ;; document created but owned by someone else
             docs2-ident2 (map (partial with-owner ident2) docs-2)
             ;; document not created
             docs3-ident1 (map (partial with-owner ident1) docs-3)
```

File: test/ctia/stores/es/crud_test.clj [hunk 8]
```diff
@@ -581,3 +607,105 @@
...
+             expected-delete-result {:deleted (set (map :id docs-1))
+                                     :errors {:not-found expected-not-found}}
+             expected-update-result {:updated (set (map :id docs-1))
+                                     :errors {:not-found expected-not-found}}]
+         (is (= expected-update-result
+                (-> (update-in update-res [:errors :not-found] set)
+                    (update :updated set))))
+         (is (= expected-delete-result
+                (-> (update-in delete-res [:errors :not-found] set)
+                    (update :deleted set)))))))))
```
Comment (yogsototh) [inline test/ctia/stores/es/crud_test.clj:711]: I think that if you want to test how `bulk-update` and `bulk-delete` behave together you probably want to test more scenarios.

1. (this one) generate n docs => check you can update then delete them all
2. generate n docs => update the 1->k and delete all even ones
3. create a doc, update it, delete it, try to re-update it, try to re-delete it
4. create a doc, concurrently update it and delete it
5. same as 2 but concurrently, would there always be half entities? Does some update could occurs concurrently and thus have any number of entites from n/2 to n? 

So you could expose some properties to provide the correct expectations (does delete enforce a 100% certainty of deletion or not).

File: src/ctia/stores/es/crud.clj [hunk 9]
```diff
@@ -92,19 +87,19 @@
...
-
-(s/defn get-docs-with-indices
+(defn get-docs-with-indices
   "Retrieves a documents from a search \"ids\" query. It enables to retrieves
  documents from an alias that points to multiple indices.
 It returns the documents with full hits meta data including the real index in which is stored the document."
-  [{:keys [conn index]} :- ESConnState
-   ids :- [s/Str]
+  [{:keys [conn index] :as _conn-state}
+   ids
```
Comment (yogsototh) [inline src/ctia/stores/es/crud.clj:95]: Why did you remove the schemas here? They looked fine.

File: src/ctia/stores/es/crud.clj [hunk 10]
```diff
@@ -117,32 +112,48 @@ It returns the documents with full hits meta data including the real index in wh
...
+         :_index (:write-index props)
+         :_type (name mapping)))
+
 (defn handle-create
   "Generate an ES create handler using some mapping and schema"
   [mapping Model]
   (let [coerce! (coerce-to-fn (s/maybe Model))]
     (s/fn :- (s/maybe [Model])
-      [{:keys [conn props] :as conn-state} :- ESConnState
+      [{:keys [conn] :as conn-state} :- ESConnState
```
Comment (yogsototh) [inline src/ctia/stores/es/crud.clj:144]: 🔴 

How can you use `:-` in a non `s/defn` declaration?

File: src/ctia/stores/es/crud.clj [hunk 11]
```diff
@@ -117,32 +112,48 @@ It returns the documents with full hits meta data including the real index in wh
...
        docs :- [Model]
        _ident
        es-params]
-      (let [prepare-doc #(assoc %
-                                :_id (:id %)
-                                :_index (:write-index props)
-                                :_type (name mapping))]
+      (let [prepare-doc (partial prepare-bulk-doc conn-state mapping)
+            prepared (doall
+                      (map prepare-doc docs))]
```
Comment (yogsototh) [inline src/ctia/stores/es/crud.clj:150]:
```suggestion
            prepared (mapv prepare-doc docs)]
```

File: src/ctia/stores/es/crud.clj [hunk 12]
```diff
@@ -117,32 +112,48 @@ It returns the documents with full hits meta data including the real index in wh
...
         (try
-          (map #(build-create-result % coerce!)
-               (ductile.doc/bulk-create-doc conn
-                                            (map prepare-doc docs)
-                                            (prepare-opts conn-state es-params)))
+          (ductile.doc/bulk-index-docs conn
+                                       prepared
+                                       (prepare-opts conn-state es-params)
+                                       50000)
```
Comment (yogsototh) [inline src/ctia/stores/es/crud.clj:155]: Why this value? Shouldn't you put this magic number in the configuration? Or perhaps get this value depending on some ES configuration value?

File: src/ctia/stores/es/crud.clj [hunk 13]
```diff
@@ -202,24 +213,125 @@ It returns the documents with full hits meta data including the real index in wh
...
+  (st/optional-keys
+   {:deleted [s/Str]
+    :updated [s/Str]
+    :errors (st/optional-keys
+             {:forbidden [s/Str]
+              :not-found [s/Str]
+              :internal-error [s/Str]})}))
+
+(defn- format-bulk-res
+  [bulk-res]
```
Comment (yogsototh) [inline src/ctia/stores/es/crud.clj:226]: This assume a certain contstrained format, for documentation sake it would probably be nice to add a few schemas to this definition.

Even nicer, I think a docstring with an example of expected value would be nice.
For example the `(group-by :result)` is not really natural to follow for me.

File: src/ctia/stores/es/crud.clj [hunk 14]
```diff
@@ -202,24 +213,125 @@ It returns the documents with full hits meta data including the real index in wh
...
+      updated (assoc :updated updated)
+      not_found (assoc-in [:errors :not-found] not_found))))
+
+(defn check-and-prepare-bulk
+  [{{{:keys [get-in-config]} :ConfigService} :services
+    :as conn-state}
+   ids
+   ident]
+  (let [doc-ids (map ensure-document-id ids)
+        docs-with-indices (get-docs-with-indices conn-state doc-ids {})
```
Comment (yogsototh) [inline src/ctia/stores/es/crud.clj:245]: generic remark, but it would be very useful to find a version of `get-docs-with-indices` that could take a ident as parameter to build an ES request that will returns only docs the ident can read (or write depending on a param).

File: src/ctia/stores/es/crud.clj [hunk 15]
```diff
@@ -202,24 +213,125 @@ It returns the documents with full hits meta data including the real index in wh
...
+             (group-by :result)
+             (into {}
+                   (map (fn [[result items]]
+                          {(keyword result) (map :_id items)}))))]
+    (cond-> {}
+      deleted (assoc :deleted deleted)
+      updated (assoc :updated updated)
+      not_found (assoc-in [:errors :not-found] not_found))))
+
+(defn check-and-prepare-bulk
```
Comment (yogsototh) [inline src/ctia/stores/es/crud.clj:239]: A short docstring could probably help here.

File: src/ctia/stores/es/crud.clj [hunk 16]
```diff
@@ -202,24 +213,125 @@ It returns the documents with full hits meta data including the real index in wh
...
+  (let [{:keys [prepared errors]} (check-and-prepare-bulk conn-state ids ident)
+        bulk-res (when prepared
+                   (try
+                     (format-bulk-res
+                      (ductile.doc/bulk-delete-docs conn
+                                                    prepared
+                                                    (prepare-opts conn-state es-params)))
+                     (catch Exception e
+                       (log/error (str "bulk delete failed: " (.getMessage e))
+                                  (pr-str prepared))
```
Comment (yogsototh) [inline src/ctia/stores/es/crud.clj:276]:
```suggestion
                       (log/error e (str "bulk delete failed: " (.getMessage e))
                                  (pr-str prepared))
```

File: src/ctia/stores/es/crud.clj [hunk 17]
```diff
@@ -202,24 +213,125 @@ It returns the documents with full hits meta data including the real index in wh
...
+          {:keys [prepared errors]} (check-and-prepare-bulk conn-state
+                                                            ids
+                                                            ident)
+
+          prepared-docs (map (fn [meta]
+                               (-> (:_id meta)
+                                   by-id
+                                   first
+                                   (into meta)))
+                             prepared)
```
Comment (yogsototh) [inline src/ctia/stores/es/crud.clj:301]: While this looks easier to read for me please check this is not a lot slower than `group-by`.
Ideally I would probably had done something along `(let [by-id (fmap first (group-by :id docs)),,,`:

```suggestion
    (let [by-id (into {} (for [{:keys [id] :as doc} docs] [id doc])
          ids (seq (keys by-id))
          {:keys [prepared errors]} (check-and-prepare-bulk conn-state
                                                            ids
                                                            ident)

          prepared-docs (map (fn [{:keys [_id] :as meta}]
                               (into (get by-id _id) meta))
                             prepared)
```

**Review by yogsototh — CHANGES_REQUESTED** [review]
I think there is a potential bug with a schema annotation on a non `s/def` function declaration.

I just made a few comments. The PR is big, I will revisit it a bit later.
Except this schema annotation I don't see any major issue.

File: src/ctia/stores/es/crud.clj (hunk 10 above)
Comment (ereteog) [reply src/ctia/stores/es/crud.clj:144]: This is `s/fn` declaration :-)
http://plumatic.github.io/schema/schema.core.html#var-fn

File: src/ctia/stores/es/crud.clj (hunk 5 above)
Comment (ereteog) [reply src/ctia/stores/es/crud.clj]: I do live in the wood, but I grew up in the hood

File: test/ctia/bulk/routes_test.clj [hunk 18]
```diff
@@ -318,7 +316,38 @@
...
+                 (DELETE app
+                         (str "ctia/bulk?"
+                              (make-get-query-str-from-bulkrefs deleted-bulk-ids))
+                         :headers {"Authorization" "45c1f5e3f05d0"})
+                 {get-res :parsed-body}
+                 (GET app
+                      (str "ctia/bulk?"
+                           (make-get-query-str-from-bulkrefs bulk-ids))
+                      :headers {"Authorization" "45c1f5e3f05d0"})]
+             (clojure.pprint/pprint res)
```
Comment (agzam) [inline test/ctia/bulk/routes_test.clj:341]: This one is here deliberately? Or you meant to remove it later?

File: test/ctia/flows/crud_test.clj [hunk 19]
```diff
@@ -122,3 +126,113 @@
...
+                                :entity-type :sighting
+                                :get-fn get-fn
+                                :delete-fn (fn [[id]]
+                                             (not= @store
+                                                   (swap! store dissoc id)))
+                                :entity-ids [entity-id]
+                                :long-id-fn identity
+                                :services services
+                                :identity (map->Identity ident))
+                               _ (Thread/sleep 1000)
```
[the remark this answers is not in the document]
Comment (ereteog) [reply test/ctia/flows/crud_test.clj:170]: I chose a different strategy, I set the default refresh to `true` in the properties.

File: test/ctia/flows/crud_test.clj [hunk 20]
```diff
@@ -122,3 +126,113 @@
...
+           store (atom {"sighting-1" (assoc sighting-minimal
+                                            :title "sighting one"
+                                            :id "sighting-1"
+                                            :tlp "green"
+                                            :groups ["groups1"])
+                        "sighting-2" (assoc sighting-minimal
+                                            :title "sighting two"
+                                            :id "sighting-2"
+                                            :tlp "green"
+                                            :groups ["groups1"])})
```
[the remark this answers is not in the document]
Comment (ereteog) [reply test/ctia/flows/crud_test.clj:153]: I made the make sighting but preserved the duplication for the ids in the map to ease the readability of the test.

File: test/ctia/flows/crud_test.clj [hunk 21]
```diff
@@ -122,3 +126,113 @@
...
+                         (let [entity-ids (seq (keys expected))
+                               res
+                               (flows.crud/delete-flow
+                                :entity-type :sighting
+                                :get-fn get-fn
+                                :delete-fn delete-fn
+                                :entity-ids entity-ids
+                                :long-id-fn identity
+                                :services services
+                                :identity (map->Identity ident))
```
[the remark this answers is not in the document]
Comment (ereteog) [reply test/ctia/flows/crud_test.clj:222]: I removed the first one which I mostly done to ease my development.

File: src/ctia/stores/es/crud.clj (hunk 9 above)
Comment (ereteog) [reply src/ctia/stores/es/crud.clj:95]: probably a trace of some dev fight with myself :-)

File: src/ctia/stores/es/crud.clj (hunk 12 above)
Comment (ereteog) [reply src/ctia/stores/es/crud.clj:155]: another trace of a dev fight. I remove it, it is solved with https://github.com/threatgrid/ductile/pull/16

File: src/ctia/stores/es/crud.clj (hunk 17 above)
Comment (ereteog) [reply src/ctia/stores/es/crud.clj:308]: group-by seems faster
``` clojure
user> (def docs (map #(array-map :id % :title (str %)) (range 1000)))
;; => #'user/docs
user> (dotimes [n 10]
         (println "(group-by :id docs)")
         (time (group-by :id docs))
         (println "(into {} (for [{:keys [id] :as doc} docs] [id doc]))")
         (time (into {} (for [{:keys [id] :as doc} docs] [id doc]))))
(group-by :id docs)
"Elapsed time: 1.010702 msecs"
(into {} (for [{:keys [id] :as doc} docs] [id doc]))
"Elapsed time: 1.520924 msecs"
(group-by :id docs)
"Elapsed time: 0.361242 msecs"
(into {} (for [{:keys [id] :as doc} docs] [id doc]))
"Elapsed time: 0.516849 msecs"
(group-by :id docs)
"Elapsed time: 0.368968 msecs"
(into {} (for [{:keys [id] :as doc} docs] [id doc]))
"Elapsed time: 0.591694 msecs"
(group-by :id docs)
"Elapsed time: 0.523959 msecs"
(into {} (for [{:keys [id] :as doc} docs] [id doc]))
"Elapsed time: 0.627433 msecs"
(group-by :id docs)
"Elapsed time: 0.384577 msecs"
(into {} (for [{:keys [id] :as doc} docs] [id doc]))
"Elapsed time: 0.735438 msecs"
(group-by :id docs)
"Elapsed time: 0.359023 msecs"
(into {} (for [{:keys [id] :as doc} docs] [id doc]))
"Elapsed time: 0.400209 msecs"
(group-by :id docs)
"Elapsed time: 0.361053 msecs"
(into {} (for [{:keys [id] :as doc} docs] [id doc]))
"Elapsed time: 0.400394 msecs"
(group-by :id docs)
"Elapsed time: 0.353588 msecs"
(into {} (for [{:keys [id] :as doc} docs] [id doc]))
"Elapsed time: 0.420125 msecs"
(group-by :id docs)
"Elapsed time: 0.354192 msecs"
(into {} (for [{:keys [id] :as doc} docs] [id doc]))
"Elapsed time: 0.390267 msecs"
(group-by :id docs)
"Elapsed time: 0.366972 msecs"
(into {} (for [{:keys [id] :as doc} docs] [id doc]))
"Elapsed time: 0.390375 msecs"
```

File: test/ctia/stores/es/crud_test.clj (hunk 8 above)
Comment (ereteog) [reply test/ctia/stores/es/crud_test.clj:714]: I added a comment in the bulk update task, I prefer to limit the inflation of this PR :-)
https://github.com/advthreat/iroh/issues/2249#issuecomment-895908314

**Review by yogsototh — APPROVED** [review]
Unblock the PR.
